### PR TITLE
Implement alliances in editor

### LINKED ISSUE #8715: Add an ability to edit alliances within the Editor
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

The original Editor allows to edit alliances only divided in between. The current one doesn't allow any.

### Describe the possible solution

We need to allow to create alliances of any combination of colors.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by Districh-ru**
Close #8715 

img

### CODE REVIEW

**Comment by Districh-ru** [comment]
Hi, @zenseii, could you please check the new phrases?

File: src/fheroes2/editor/editor_map_specs_window.cpp [hunk 1]
```diff
@@ -486,7 +486,29 @@ namespace
...
-        return nullptr;
+        return "Unknown";
+    }
+
+    // Returns the order text of the alliance: 0 will be "1st" and so on.
+    const char * getAllianceNumberText( const size_t index )
+    {
+        switch ( index ) {
+        case 0:
+            return _( "1st" );
```
Comment (ihhub) [inline src/fheroes2/editor/editor_map_specs_window.cpp:497]: Some languages support different forms of adjective depending on a noun. Also, we have almost the same function in the code `const char * getPlayerOrderString( const size_t player )`.  I suggest to add a special word before each of this entry. For example, `alliance|1st`. In this case translations are going to be more natural. The same should be done for the mentioned function.

What do you think @zenseii ?

File: src/fheroes2/editor/editor_map_specs_window.cpp [hunk 2]
```diff
@@ -1291,6 +1422,7 @@ namespace
 
     private:
         uint8_t _conditionType{ Maps::FileInfo::VICTORY_DEFEAT_EVERYONE };
+        uint8_t _availableColors{ Color::NONE };
```
Comment (ihhub) [inline src/fheroes2/editor/editor_map_specs_window.cpp:1425]: I think this variable can be `const`.

File: src/fheroes2/editor/editor_map_specs_window.cpp [hunk 3]
```diff
@@ -486,7 +486,29 @@ namespace
             break;
         }
 
-        return nullptr;
+        return "Unknown";
+    }
+
+    // Returns the order text of the alliance: 0 will be "1st" and so on.
+    const char * getAllianceNumberText( const size_t index )
```
Comment (ihhub) [inline src/fheroes2/editor/editor_map_specs_window.cpp:493]: The more appropriate name for this function could be `getAllianceOrderText` as number is logically 1, 2, 3 and etc.

File: src/fheroes2/editor/editor_map_specs_window.cpp [hunk 4]
```diff
@@ -739,6 +755,21 @@ namespace
...
+            case Maps::FileInfo::VICTORY_DEFEAT_OTHER_SIDE:
+                // As of now only 2 alliances are supported.
+                // TODO: Update this logic for more than 2 alliances.
+                if ( mapFormat.alliances.size() == 2 && ( mapFormat.alliances[0] & mapFormat.alliances[1] ) == 0 ) {
+                    const int playersCount = Color::Count( mapFormat.availablePlayerColors );
+                    const int alliance1Count = Color::Count( mapFormat.alliances[0] );
+                    const int alliance2Count = Color::Count( mapFormat.alliances[1] );
+
+                    // Use the saved in mapFormat alliances only if they are correct.
+                    if ( playersCount != alliance1Count && playersCount != alliance2Count && alliance1Count != 0 && alliance2Count != 0 ) {
```
Comment (ihhub) [inline src/fheroes2/editor/editor_map_specs_window.cpp:768]: We should have the following checks:
```cpp
( mapFormat.alliances[0] & mapFormat.alliances[1] ) == 0 // to make sure that no alliances have the same colors
mapFormat.availablePlayerColors == ( mapFormat.alliances[0] | mapFormat.alliances[1] ) // to make sure that alliances contain all colors
mapFormat.alliances[0] != 0 // to check that the alliance is not empty
mapFormat.alliances[1] != 0 // to check that the alliance is not empty
```
so, the above variables are not needed.

File: src/fheroes2/editor/editor_map_specs_window.cpp [hunk 5]
```diff
@@ -873,6 +913,17 @@ namespace
                 mapFormat.victoryConditionMetadata[0] = ( _victoryArtifactId == ultimateArtifactId ) ? 0 : _victoryArtifactId;
                 mapFormat.allowNormalVictory = _isNormalVictoryAllowed;
 
+                return;
+            case Maps::FileInfo::VICTORY_DEFEAT_OTHER_SIDE:
+                if ( mapFormat.alliances.size() != 2 ) {
+                    mapFormat.alliances.resize( 2 );
+                }
+
+                std::copy( _alliances.begin(), _alliances.end(), mapFormat.alliances.begin() );
```
Comment (ihhub) [inline src/fheroes2/editor/editor_map_specs_window.cpp:922]: I think we can use a shorter code:
```cpp
mapFormat.alliances = _alliances;
```

**Review by ihhub — APPROVED** [review]

**Comment by ihhub** [comment]
@Districh-ru , thank you so much for this feature!

**Pull request merged**